README.md
Yuyo GTK Theme
================

Yuyo is an GTK theme belonging to the [Ubuntu MATE](http://www.ubuntu-mate.org/) project, it is a fork of the Orchis GTK theme from [Moka Project](http://mokaproject.com/).

Yuyo GTK Theme is distributed under the terms the GNU GPL v.3

###Preamble

If you find any bugs or issues with Yuyo or if you have a question, you can visit Yuyo's primary issue tracker on [GitHub](https://github.com/snwh/yuyo-gtk-theme/issues).

###Getting the Source

The source for Yuyo GTK Theme can be found [here](https://github.com/snwh/yuyo-gtk-theme).

Alternatively, you can clone the latest version its git repository:

    git clone https://github.com/snwh/yuyo-gtk-theme.git

###Using the Source

There are scripts to simplify the rendering process;to run them (and edit icons) you will need:

 * inkscape
 * python3

To render new assets from their source SVG files, either run the following:

    ./render-gtk3-assets.py
    ./render-gtk3-assets-hidpi.py
    ./render-metacity-assets.py
    ./render-unity-assets.py

Other asset sources are hidden a bit deeper. For borders:

    cd src/ 
    ./render-borders.sh

And GTK2 assets: 

    cd src/gtk2
    ./render-bits.sh

If it's throwing an error, the script may not be executable, try:
	
	chmod +x *

This script will look in the source directories (../src/*) and render the respective icons (provided there are changes).

-----------

debian/changelog
yuyo-gtk-theme (0.3) vivid; urgency=low

  * Override symbolic icons.
  * Move window buttons 1px up.
  * Closes (LP: #1427189)

 -- Martin Wimpress <code@flexion.org>  Sun, 08 Feb 2014 22:48:21 -0000

yuyo-gtk-theme (0.2) vivid; urgency=low

  * Updated packaging.

 -- Martin Wimpress <code@flexion.org>  Wed, 04 Jan 2014 00:51:00 -0000

yuyo-gtk-theme (0.1) vivid; urgency=low

  * Initial Release.

 -- Martin Wimpress <code@flexion.org>  Wed, 10 Dec 2014 22:03:00 -0000


render-gtk3-assets.py
#!/usr/bin/python3

import os
import sys
import xml.sax
import subprocess

INKSCAPE = '/usr/bin/inkscape'
OPTIPNG = '/usr/bin/optipng'
SRC = os.path.join('.', 'src/gtk3')

inkscape_process = None

def optimize_png(png_file):
    if os.path.exists(OPTIPNG):
        process = subprocess.Popen([OPTIPNG, '-quiet', '-o7', png_file])
        process.wait()

def wait_for_prompt(process, command=None):
    if command is not None:
        process.stdin.write((command+'\n').encode('utf-8'))

    # This is kinda ugly ...
    # Wait for just a '>', or '\n>' if some other char appearead first
    output = process.stdout.read(1)
    if output == b'>':
        return

    output += process.stdout.read(1)
    while output != b'\n>':
        output += process.stdout.read(1)
        output = output[1:]

def start_inkscape():
    process = subprocess.Popen([INKSCAPE, '--shell'], bufsize=0, stdin=subprocess.PIPE, stdout=subprocess.PIPE)
    wait_for_prompt(process)
    return process

def inkscape_render_rect(icon_file, rect, output_file):
    global inkscape_process
    if inkscape_process is None:
        inkscape_process = start_inkscape()
    wait_for_prompt(inkscape_process, '%s -i %s -e %s' % (icon_file, rect, output_file))
    optimize_png(output_file)

class ContentHandler(xml.sax.ContentHandler):
    ROOT = 0
    SVG = 1
    LAYER = 2
    OTHER = 3
    TEXT = 4
    def __init__(self, path, force=False, filter=None):
        self.stack = [self.ROOT]
        self.inside = [self.ROOT]
        self.path = path
        self.rects = []
        self.state = self.ROOT
        self.chars = ""
        self.force = force
        self.filter = filter

    def endDocument(self):
        pass

    def startElement(self, name, attrs):
        if self.inside[-1] == self.ROOT:
            if name == "svg":
                self.stack.append(self.SVG)
                self.inside.append(self.SVG)
                return
        elif self.inside[-1] == self.SVG:
            if (name == "g" and ('inkscape:groupmode' in attrs) and ('inkscape:label' in attrs)
               and attrs['inkscape:groupmode'] == 'layer' and attrs['inkscape:label'].startswith('Baseplate')):
                self.stack.append(self.LAYER)
                self.inside.append(self.LAYER)
                self.context = None
                self.icon_name = None
                self.rects = []
                return
        elif self.inside[-1] == self.LAYER:
            if name == "text" and ('inkscape:label' in attrs) and attrs['inkscape:label'] == 'context':
                self.stack.append(self.TEXT)
                self.inside.append(self.TEXT)
                self.text='context'
                self.chars = ""
                return
            elif name == "text" and ('inkscape:label' in attrs) and attrs['inkscape:label'] == 'icon-name':
                self.stack.append(self.TEXT)
                self.inside.append(self.TEXT)
                self.text='icon-name'
                self.chars = ""
                return
            elif name == "rect":
                self.rects.append(attrs)

        self.stack.append(self.OTHER)


    def endElement(self, name):
        stacked = self.stack.pop()
        if self.inside[-1] == stacked:
            self.inside.pop()

        if stacked == self.TEXT and self.text is not None:
            assert self.text in ['context', 'icon-name']
            if self.text == 'context':
                self.context = self.chars
            elif self.text == 'icon-name':
                self.icon_name = self.chars
            self.text = None
        elif stacked == self.LAYER:
            assert self.icon_name
            assert self.context

            if self.filter is not None and not self.icon_name in self.filter:
                return

            print (self.context, self.icon_name)
            for rect in self.rects:
                width = rect['width']
                height = rect['height']
                id = rect['id']

                dir = os.path.join("Yuyo", "gtk-3.0", self.context)
                outfile = os.path.join(dir, self.icon_name+'.png')
                if not os.path.exists(dir):
                    os.makedirs(dir)
                # Do a time based check!
                if self.force or not os.path.exists(outfile):
                    inkscape_render_rect(self.path, id, outfile)
                    sys.stdout.write('.')
                else:
                    stat_in = os.stat(self.path)
                    stat_out = os.stat(outfile)
                    if stat_in.st_mtime > stat_out.st_mtime:
                        inkscape_render_rect(self.path, id, outfile)
                        sys.stdout.write('.')
                    else:
                        sys.stdout.write('-')
                sys.stdout.flush()
            sys.stdout.write('\n')
            sys.stdout.flush()

    def characters(self, chars):
        self.chars += chars.strip()

if len(sys.argv) == 1:
    if not os.path.exists('Yuyo'):
        os.mkdir('Yuyo')
    print ('Rendering from SVGs in', SRC)
    for file in os.listdir(SRC):
        if file[-4:] == '.svg':
            file = os.path.join(SRC, file)
            handler = ContentHandler(file)
            xml.sax.parse(open(file), handler)
else:
    file = os.path.join(SRC, sys.argv[1] + '.svg')
    if len(sys.argv) > 2:
        icons = sys.argv[2:]
    else:
        icons = None
    if os.path.exists(os.path.join(file)):
        handler = ContentHandler(file, True, filter=icons)
        xml.sax.parse(open(file), handler)
    else:
        print ("Error: No such file", file)
        sys.exit(1)




src/render-borders.sh
#! /bin/bash

INKSCAPE="/usr/bin/inkscape"
OPTIPNG="/usr/bin/optipng"

SRC_FILE="gtk3-borders.svg"
ASSETS_DIR="borders"
INDEX="borders.txt"

for i in `cat $INDEX`
do 
if [ -f $ASSETS_DIR/$i.png ]; then
    echo $ASSETS_DIR/$i.png exists.
else
    echo
    echo Rendering $ASSETS_DIR/$i.png
    $INKSCAPE --export-id=$i \
              --export-id-only \
              --export-png=$ASSETS_DIR/$i.png $SRC_FILE >/dev/null #\
    # && $OPTIPNG -o7 --quiet $ASSETS_DIR/$i.png 
fi
if [ -f $ASSETS_DIR/$i@2.png ]; then
    echo $ASSETS_DIR/$i@2.png exists.
else
    echo
    echo Rendering $ASSETS_DIR/$i@2.png
    $INKSCAPE --export-id=$i \
              --export-dpi=180 \
              --export-id-only \
              --export-png=$ASSETS_DIR/$i@2.png $SRC_FILE >/dev/null #\
    # && $OPTIPNG -o7 --quiet $ASSETS_DIR/$i@2.png 
fi
done
exit 0


render-unity-assets.py
#!/usr/bin/python3

import os
import sys
import xml.sax
import subprocess

INKSCAPE = '/usr/bin/inkscape'
OPTIPNG = '/usr/bin/optipng'
SRC = os.path.join('.', 'src/unity')

inkscape_process = None

def optimize_png(png_file):
    if os.path.exists(OPTIPNG):
        process = subprocess.Popen([OPTIPNG, '-quiet', '-o7', png_file])
        process.wait()

def wait_for_prompt(process, command=None):
    if command is not None:
        process.stdin.write((command+'\n').encode('utf-8'))

    # This is kinda ugly ...
    # Wait for just a '>', or '\n>' if some other char appearead first
    output = process.stdout.read(1)
    if output == b'>':
        return

    output += process.stdout.read(1)
    while output != b'\n>':
        output += process.stdout.read(1)
        output = output[1:]

def start_inkscape():
    process = subprocess.Popen([INKSCAPE, '--shell'], bufsize=0, stdin=subprocess.PIPE, stdout=subprocess.PIPE)
    wait_for_prompt(process)
    return process

def inkscape_render_rect(icon_file, rect, output_file):
    global inkscape_process
    if inkscape_process is None:
        inkscape_process = start_inkscape()
    wait_for_prompt(inkscape_process, '%s -i %s -e %s' % (icon_file, rect, output_file))
    optimize_png(output_file)

class ContentHandler(xml.sax.ContentHandler):
    ROOT = 0
    SVG = 1
    LAYER = 2
    OTHER = 3
    TEXT = 4
    def __init__(self, path, force=False, filter=None):
        self.stack = [self.ROOT]
        self.inside = [self.ROOT]
        self.path = path
        self.rects = []
        self.state = self.ROOT
        self.chars = ""
        self.force = force
        self.filter = filter

    def endDocument(self):
        pass

    def startElement(self, name, attrs):
        if self.inside[-1] == self.ROOT:
            if name == "svg":
                self.stack.append(self.SVG)
                self.inside.append(self.SVG)
                return
        elif self.inside[-1] == self.SVG:
            if (name == "g" and ('inkscape:groupmode' in attrs) and ('inkscape:label' in attrs)
               and attrs['inkscape:groupmode'] == 'layer' and attrs['inkscape:label'].startswith('Baseplate')):
                self.stack.append(self.LAYER)
                self.inside.append(self.LAYER)
                self.context = None
                self.icon_name = None
                self.rects = []
                return
        elif self.inside[-1] == self.LAYER:
            if name == "text" and ('inkscape:label' in attrs) and attrs['inkscape:label'] == 'context':
                self.stack.append(self.TEXT)
                self.inside.append(self.TEXT)
                self.text='context'
                self.chars = ""
                return
            elif name == "text" and ('inkscape:label' in attrs) and attrs['inkscape:label'] == 'icon-name':
                self.stack.append(self.TEXT)
                self.inside.append(self.TEXT)
                self.text='icon-name'
                self.chars = ""
                return
            elif name == "rect":
                self.rects.append(attrs)

        self.stack.append(self.OTHER)


    def endElement(self, name):
        stacked = self.stack.pop()
        if self.inside[-1] == stacked:
            self.inside.pop()

        if stacked == self.TEXT and self.text is not None:
            assert self.text in ['context', 'icon-name']
            if self.text == 'context':
                self.context = self.chars
            elif self.text == 'icon-name':
                self.icon_name = self.chars
            self.text = None
        elif stacked == self.LAYER:
            assert self.icon_name
            assert self.context

            if self.filter is not None and not self.icon_name in self.filter:
                return

            print (self.context, self.icon_name)
            for rect in self.rects:
                width = rect['width']
                height = rect['height']
                id = rect['id']

                dir = os.path.join("Yuyo", self.context)
                outfile = os.path.join(dir, self.icon_name+'.png')
                if not os.path.exists(dir):
                    os.makedirs(dir)
                # Do a time based check!
                if self.force or not os.path.exists(outfile):
                    inkscape_render_rect(self.path, id, outfile)
                    sys.stdout.write('.')
                else:
                    stat_in = os.stat(self.path)
                    stat_out = os.stat(outfile)
                    if stat_in.st_mtime > stat_out.st_mtime:
                        inkscape_render_rect(self.path, id, outfile)
                        sys.stdout.write('.')
                    else:
                        sys.stdout.write('-')
                sys.stdout.flush()
            sys.stdout.write('\n')
            sys.stdout.flush()

    def characters(self, chars):
        self.chars += chars.strip()

if len(sys.argv) == 1:
    if not os.path.exists('Yuyo'):
        os.mkdir('Yuyo')
    print ('Rendering from SVGs in', SRC)
    for file in os.listdir(SRC):
        if file[-4:] == '.svg':
            file = os.path.join(SRC, file)
            handler = ContentHandler(file)
            xml.sax.parse(open(file), handler)
else:
    file = os.path.join(SRC, sys.argv[1] + '.svg')
    if len(sys.argv) > 2:
        icons = sys.argv[2:]
    else:
        icons = None
    if os.path.exists(os.path.join(file)):
        handler = ContentHandler(file, True, filter=icons)
        xml.sax.parse(open(file), handler)
    else:
        print ("Error: No such file", file)
        sys.exit(1)




render-metacity-assets.py
#!/usr/bin/python3

import os
import sys
import xml.sax
import subprocess

INKSCAPE = '/usr/bin/inkscape'
OPTIPNG = '/usr/bin/optipng'
SRC = os.path.join('.', 'src/metacity')

inkscape_process = None

def optimize_png(png_file):
    if os.path.exists(OPTIPNG):
        process = subprocess.Popen([OPTIPNG, '-quiet', '-o7', png_file])
        process.wait()

def wait_for_prompt(process, command=None):
    if command is not None:
        process.stdin.write((command+'\n').encode('utf-8'))

    # This is kinda ugly ...
    # Wait for just a '>', or '\n>' if some other char appearead first
    output = process.stdout.read(1)
    if output == b'>':
        return

    output += process.stdout.read(1)
    while output != b'\n>':
        output += process.stdout.read(1)
        output = output[1:]

def start_inkscape():
    process = subprocess.Popen([INKSCAPE, '--shell'], bufsize=0, stdin=subprocess.PIPE, stdout=subprocess.PIPE)
    wait_for_prompt(process)
    return process

def inkscape_render_rect(icon_file, rect, output_file):
    global inkscape_process
    if inkscape_process is None:
        inkscape_process = start_inkscape()
    wait_for_prompt(inkscape_process, '%s -i %s -e %s' % (icon_file, rect, output_file))
    optimize_png(output_file)

class ContentHandler(xml.sax.ContentHandler):
    ROOT = 0
    SVG = 1
    LAYER = 2
    OTHER = 3
    TEXT = 4
    def __init__(self, path, force=False, filter=None):
        self.stack = [self.ROOT]
        self.inside = [self.ROOT]
        self.path = path
        self.rects = []
        self.state = self.ROOT
        self.chars = ""
        self.force = force
        self.filter = filter

    def endDocument(self):
        pass

    def startElement(self, name, attrs):
        if self.inside[-1] == self.ROOT:
            if name == "svg":
                self.stack.append(self.SVG)
                self.inside.append(self.SVG)
                return
        elif self.inside[-1] == self.SVG:
            if (name == "g" and ('inkscape:groupmode' in attrs) and ('inkscape:label' in attrs)
               and attrs['inkscape:groupmode'] == 'layer' and attrs['inkscape:label'].startswith('Baseplate')):
                self.stack.append(self.LAYER)
                self.inside.append(self.LAYER)
                self.context = None
                self.icon_name = None
                self.rects = []
                return
        elif self.inside[-1] == self.LAYER:
            if name == "text" and ('inkscape:label' in attrs) and attrs['inkscape:label'] == 'context':
                self.stack.append(self.TEXT)
                self.inside.append(self.TEXT)
                self.text='context'
                self.chars = ""
                return
            elif name == "text" and ('inkscape:label' in attrs) and attrs['inkscape:label'] == 'icon-name':
                self.stack.append(self.TEXT)
                self.inside.append(self.TEXT)
                self.text='icon-name'
                self.chars = ""
                return
            elif name == "rect":
                self.rects.append(attrs)

        self.stack.append(self.OTHER)


    def endElement(self, name):
        stacked = self.stack.pop()
        if self.inside[-1] == stacked:
            self.inside.pop()

        if stacked == self.TEXT and self.text is not None:
            assert self.text in ['context', 'icon-name']
            if self.text == 'context':
                self.context = self.chars
            elif self.text == 'icon-name':
                self.icon_name = self.chars
            self.text = None
        elif stacked == self.LAYER:
            assert self.icon_name
            assert self.context

            if self.filter is not None and not self.icon_name in self.filter:
                return

            print (self.context, self.icon_name)
            for rect in self.rects:
                width = rect['width']
                height = rect['height']
                id = rect['id']

                dir = os.path.join("Yuyo", self.context)
                outfile = os.path.join(dir, self.icon_name+'.png')
                if not os.path.exists(dir):
                    os.makedirs(dir)
                # Do a time based check!
                if self.force or not os.path.exists(outfile):
                    inkscape_render_rect(self.path, id, outfile)
                    sys.stdout.write('.')
                else:
                    stat_in = os.stat(self.path)
                    stat_out = os.stat(outfile)
                    if stat_in.st_mtime > stat_out.st_mtime:
                        inkscape_render_rect(self.path, id, outfile)
                        sys.stdout.write('.')
                    else:
                        sys.stdout.write('-')
                sys.stdout.flush()
            sys.stdout.write('\n')
            sys.stdout.flush()

    def characters(self, chars):
        self.chars += chars.strip()

if len(sys.argv) == 1:
    if not os.path.exists('Yuyo'):
        os.mkdir('Yuyo')
    print ('Rendering from SVGs in', SRC)
    for file in os.listdir(SRC):
        if file[-4:] == '.svg':
            file = os.path.join(SRC, file)
            handler = ContentHandler(file)
            xml.sax.parse(open(file), handler)
else:
    file = os.path.join(SRC, sys.argv[1] + '.svg')
    if len(sys.argv) > 2:
        icons = sys.argv[2:]
    else:
        icons = None
    if os.path.exists(os.path.join(file)):
        handler = ContentHandler(file, True, filter=icons)
        xml.sax.parse(open(file), handler)
    else:
        print ("Error: No such file", file)
        sys.exit(1)




render-gtk3-assets-hidpi.py
#!/usr/bin/python3

import os
import sys
import xml.sax
import subprocess

INKSCAPE = '/usr/bin/inkscape'
OPTIPNG = '/usr/bin/optipng'
SRC = os.path.join('.', 'src/gtk3')

inkscape_process = None

def optimize_png(png_file):
    if os.path.exists(OPTIPNG):
        process = subprocess.Popen([OPTIPNG, '-quiet', '-o7', png_file])
        process.wait()

def wait_for_prompt(process, command=None):
    if command is not None:
        process.stdin.write((command+'\n').encode('utf-8'))

    # This is kinda ugly ...
    # Wait for just a '>', or '\n>' if some other char appearead first
    output = process.stdout.read(1)
    if output == b'>':
        return

    output += process.stdout.read(1)
    while output != b'\n>':
        output += process.stdout.read(1)
        output = output[1:]

def start_inkscape():
    process = subprocess.Popen([INKSCAPE, '--shell'], bufsize=0, stdin=subprocess.PIPE, stdout=subprocess.PIPE)
    wait_for_prompt(process)
    return process

def inkscape_render_rect(icon_file, rect, output_file):
    global inkscape_process
    if inkscape_process is None:
        inkscape_process = start_inkscape()
    wait_for_prompt(inkscape_process, '--export-dpi=180 %s -i %s -e %s' % (icon_file, rect, output_file))
    optimize_png(output_file)

class ContentHandler(xml.sax.ContentHandler):
    ROOT = 0
    SVG = 1
    LAYER = 2
    OTHER = 3
    TEXT = 4
    def __init__(self, path, force=False, filter=None):
        self.stack = [self.ROOT]
        self.inside = [self.ROOT]
        self.path = path
        self.rects = []
        self.state = self.ROOT
        self.chars = ""
        self.force = force
        self.filter = filter

    def endDocument(self):
        pass

    def startElement(self, name, attrs):
        if self.inside[-1] == self.ROOT:
            if name == "svg":
                self.stack.append(self.SVG)
                self.inside.append(self.SVG)
                return
        elif self.inside[-1] == self.SVG:
            if (name == "g" and ('inkscape:groupmode' in attrs) and ('inkscape:label' in attrs)
               and attrs['inkscape:groupmode'] == 'layer' and attrs['inkscape:label'].startswith('Baseplate')):
                self.stack.append(self.LAYER)
                self.inside.append(self.LAYER)
                self.context = None
                self.icon_name = None
                self.rects = []
                return
        elif self.inside[-1] == self.LAYER:
            if name == "text" and ('inkscape:label' in attrs) and attrs['inkscape:label'] == 'context':
                self.stack.append(self.TEXT)
                self.inside.append(self.TEXT)
                self.text='context'
                self.chars = ""
                return
            elif name == "text" and ('inkscape:label' in attrs) and attrs['inkscape:label'] == 'icon-name':
                self.stack.append(self.TEXT)
                self.inside.append(self.TEXT)
                self.text='icon-name'
                self.chars = ""
                return
            elif name == "rect":
                self.rects.append(attrs)

        self.stack.append(self.OTHER)


    def endElement(self, name):
        stacked = self.stack.pop()
        if self.inside[-1] == stacked:
            self.inside.pop()

        if stacked == self.TEXT and self.text is not None:
            assert self.text in ['context', 'icon-name']
            if self.text == 'context':
                self.context = self.chars
            elif self.text == 'icon-name':
                self.icon_name = self.chars
            self.text = None
        elif stacked == self.LAYER:
            assert self.icon_name
            assert self.context

            if self.filter is not None and not self.icon_name in self.filter:
                return

            print (self.context, self.icon_name)
            for rect in self.rects:
                width = rect['width']
                height = rect['height']
                id = rect['id']

                dir = os.path.join("Yuyo", "gtk-3.0", self.context)
                outfile = os.path.join(dir, self.icon_name+'@2'+'.png')
                if not os.path.exists(dir):
                    os.makedirs(dir)
                # Do a time based check!
                if self.force or not os.path.exists(outfile):
                    inkscape_render_rect(self.path, id, outfile)
                    sys.stdout.write('.')
                else:
                    stat_in = os.stat(self.path)
                    stat_out = os.stat(outfile)
                    if stat_in.st_mtime > stat_out.st_mtime:
                        inkscape_render_rect(self.path, id, outfile)
                        sys.stdout.write('.')
                    else:
                        sys.stdout.write('-')
                sys.stdout.flush()
            sys.stdout.write('\n')
            sys.stdout.flush()

    def characters(self, chars):
        self.chars += chars.strip()

if len(sys.argv) == 1:
    if not os.path.exists('Yuyo'):
        os.mkdir('Yuyo')
    print ('Rendering from SVGs in', SRC)
    for file in os.listdir(SRC):
        if file[-4:] == '.svg':
            file = os.path.join(SRC, file)
            handler = ContentHandler(file)
            xml.sax.parse(open(file), handler)
else:
    file = os.path.join(SRC, sys.argv[1] + '.svg')
    if len(sys.argv) > 2:
        icons = sys.argv[2:]
    else:
        icons = None
    if os.path.exists(os.path.join(file)):
        handler = ContentHandler(file, True, filter=icons)
        xml.sax.parse(open(file), handler)
    else:
        print ("Error: No such file", file)
        sys.exit(1)


